File: http/http-client-annotation-processor/src/test/java/ru/tinkoff/kora/http/client/annotation/processor/HttpClientResponseTest.java
package ru.tinkoff.kora.http.client.annotation.processor;

import org.junit.jupiter.api.Test;
import reactor.core.publisher.Mono;
import ru.tinkoff.kora.http.client.common.HttpClientResponseException;
import ru.tinkoff.kora.http.client.common.response.HttpClientResponseMapper;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class HttpClientResponseTest extends AbstractHttpClientTest {
    @Test
    public void testSimple() {
        var mapper = mock(HttpClientResponseMapper.class);
        compileClient(List.of(mapper), """
            @HttpClient
            public interface TestClient {
              @HttpRoute(method = "GET", path = "/test")
              String test();
            }
            """);
        when(mapper.apply(any())).thenReturn(Mono.just("test-string"));

        var result = client.invoke("test");

        assertThat(result).isEqualTo("test-string");
    }

    @Test
    public void testCustomFinalMapper() {
        compileClient(List.of(), """
            @HttpClient
            public interface TestClient {
              @Mapping(TestMapper.class)
              @HttpRoute(method = "GET", path = "/test")
              String test();
            }
            """, """
            public final class TestMapper implements HttpClientResponseMapper<String, Mono<String>> {
              public Mono<String> apply(HttpClientResponse rs) {
                  return Mono.just("test-string-from-mapper");
              }
            }
            """);

        var result = client.invoke("test");

        assertThat(result).isEqualTo("test-string-from-mapper");
    }

    @Test
    public void testCustomMapper() {
        compileClient(List.of(newGeneratedObject("TestMapper")), """
            @HttpClient
            public interface TestClient {
              @Mapping(TestMapper.class)
              @HttpRoute(method = "GET", path = "/test")
              String test();
            }
            """, """
            public class TestMapper implements HttpClientResponseMapper<String, Mono<String>> {
              public Mono<String> apply(HttpClientResponse rs) {
                  return Mono.just("test-string-from-mapper");
              }
            }
            """);

        var result = client.invoke("test");

        assertThat(result).isEqualTo("test-string-from-mapper");
    }

    @Test
    public void testFinalCodeMapper() {
        compileClient(List.of(), """
            @HttpClient
            public interface TestClient {
              @ResponseCodeMapper(code = 500, mapper = TestMapper.class)
              @HttpRoute(method = "GET", path = "/test")
              String test();
            }
            """, """
            public final class TestMapper implements HttpClientResponseMapper<String, Mono<String>> {
              public Mono<String> apply(HttpClientResponse rs) {
                  return Mono.just("test-string-from-mapper");
              }
            }
            """);

        when(httpResponse.code()).thenReturn(500);
        var result = client.invoke("test");
        assertThat(result).isEqualTo("test-string-from-mapper");

        when(httpResponse.code()).thenReturn(200);
        assertThatThrownBy(() -> client.invoke("test")).isInstanceOf(HttpClientResponseException.class);
    }

    @Test
    public void testCodeMapper() {
        compileClient(List.of(newGeneratedObject("TestMapper")), """
            @HttpClient
            public interface TestClient {
              @ResponseCodeMapper(code = 500, mapper = TestMapper.class)
              @HttpRoute(method = "GET", path = "/test")
              String test();
            }
            """, """
            public class TestMapper implements HttpClientResponseMapper<String, Mono<String>> {
              public Mono<String> apply(HttpClientResponse rs) {
                  return Mono.just("test-string-from-mapper");
              }
            }
            """);

        when(httpResponse.code()).thenReturn(500);
        var result = client.invoke("test");
        assertThat(result).isEqualTo("test-string-from-mapper");

        when(httpResponse.code()).thenReturn(200);
        assertThatThrownBy(() -> client.invoke("test")).isInstanceOf(HttpClientResponseException.class);
    }

    @Test
    public void testCodeMapperByType() {
        compileClient(List.of(newGeneratedObject("Test200Mapper"), newGeneratedObject("Test500Mapper")), """
            @HttpClient
            public interface TestClient {
              @ResponseCodeMapper(code = 200, type = TestResponse.Rs200.class)
              @ResponseCodeMapper(code = 500, type = TestResponse.Rs500.class)
              @HttpRoute(method = "GET", path = "/test")
              TestResponse test();
            }
            """, """
            public class Test200Mapper implements HttpClientResponseMapper<TestResponse.Rs200, Mono<TestResponse.Rs200>> {
              public Mono<TestResponse.Rs200> apply(HttpClientResponse rs) {
                  return Mono.just(new TestResponse.Rs200());
              }
            }
            """, """
            public class Test500Mapper implements HttpClientResponseMapper<TestResponse.Rs500, Mono<TestResponse.Rs500>> {
              public Mono<TestResponse.Rs500> apply(HttpClientResponse rs) {
                  return Mono.just(new TestResponse.Rs500());
              }
            }
            """, """
            public sealed interface TestResponse {
              record Rs200() implements TestResponse {}
              record Rs500() implements TestResponse {}
            }
            """);

        when(httpResponse.code()).thenReturn(500);
        var result = client.invoke("test");
        assertThat(result).isEqualTo(newObject("TestResponse$Rs500"));

        when(httpResponse.code()).thenReturn(200);
        result = client.invoke("test");
        assertThat(result).isEqualTo(newObject("TestResponse$Rs200"));

        when(httpResponse.code()).thenReturn(201);
        assertThatThrownBy(() -> client.invoke("test")).isInstanceOf(HttpClientResponseException.class);
    }
}

File: http/http-client-symbol-processor/src/test/kotlin/ru/tinkoff/kora/http/client/symbol/processor/HttpClientResponseTest.kt
package ru.tinkoff.kora.http.client.symbol.processor

import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.Test
import org.mockito.kotlin.any
import org.mockito.kotlin.mock
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever
import reactor.core.publisher.Mono
import ru.tinkoff.kora.http.client.common.HttpClientResponseException
import ru.tinkoff.kora.http.client.common.response.HttpClientResponseMapper

class HttpClientResponseTest : AbstractHttpClientTest() {
    @Test
    fun testSimple() {
        val mapper = mock<HttpClientResponseMapper<String, Mono<String>>>()
        compile(listOf(mapper), """
            @HttpClient
            interface TestClient {
              @HttpRoute(method = "GET", path = "/test")
              fun test(): String
            }
        """.trimIndent())
        whenever(mapper.apply(any())).thenReturn(Mono.just("test-string"))

        val result = client.invoke("test")

        assertThat(result).isEqualTo("test-string")
        verify(mapper).apply(any())
    }

    @Test
    fun testCustomMapper() {
        compile(listOf<Any>(), """
            @HttpClient
            interface TestClient {
              @Mapping(TestMapper::class)
              @HttpRoute(method = "GET", path = "/test")
              fun test(): String
            }
        """.trimIndent(), """
            class TestMapper : HttpClientResponseMapper<String, Mono<String>> {
              override fun apply(rs: HttpClientResponse) = Mono.just("test-string-from-mapper")!!
            }
        """.trimIndent())

        val result = client.invoke("test")

        assertThat(result).isEqualTo("test-string-from-mapper")
    }

    @Test
    fun testOpenCustomMapper() {
        compile(listOf(newGenerated("TestMapper")), """
            @HttpClient
            interface TestClient {
              @Mapping(TestMapper::class)
              @HttpRoute(method = "GET", path = "/test")
              fun test(): String
            }
        """.trimIndent(), """
            open class TestMapper : HttpClientResponseMapper<String, Mono<String>> {
              override fun apply(rs: HttpClientResponse) = Mono.just("test-string-from-mapper")!!
            }
        """.trimIndent())

        val result = client.invoke("test")

        assertThat(result).isEqualTo("test-string-from-mapper")
    }

    @Test
    fun testCodeMapperByMapper() {
        compile(listOf<Any>(), """
            @HttpClient
            interface TestClient {
              @ResponseCodeMapper(code = 500, mapper = TestMapper::class)
              @HttpRoute(method = "GET", path = "/test")
              fun test(): String
            }
        """.trimIndent(), """
            class TestMapper : HttpClientResponseMapper<String, Mono<String>> {
              override fun apply(rs: HttpClientResponse) = Mono.just("test-string-from-mapper")!!
            }
        """.trimIndent())
        whenever(httpResponse.code()).thenReturn(500)

        val result = client.invoke("test")

        assertThat(result).isEqualTo("test-string-from-mapper")

        whenever(httpResponse.code()).thenReturn(200)
        assertThatThrownBy { client.invoke("test") }.isInstanceOf(HttpClientResponseException::class.java)
    }

    @Test
    fun testOpenCodeMapperByMapper() {
        compile(listOf(newGenerated("TestMapper")), """
            @HttpClient
            interface TestClient {
              @ResponseCodeMapper(code = 500, mapper = TestMapper::class)
              @HttpRoute(method = "GET", path = "/test")
              fun test(): String
            }
        """.trimIndent(), """
            open class TestMapper : HttpClientResponseMapper<String, Mono<String>> {
              override fun apply(rs: HttpClientResponse) = Mono.just("test-string-from-mapper")!!
            }
        """.trimIndent())
        whenever(httpResponse.code()).thenReturn(500)

        val result = client.invoke("test")

        assertThat(result).isEqualTo("test-string-from-mapper")

        whenever(httpResponse.code()).thenReturn(200)
        assertThatThrownBy { client.invoke("test") }.isInstanceOf(HttpClientResponseException::class.java)
    }

    @Test
    fun testCodeMapperByType() {
        compile(listOf(newGenerated("TestMapper200"), newGenerated("TestMapper500")), """
            @HttpClient
            interface TestClient {
              @ResponseCodeMapper(code = 200, type = TestRs200::class)
              @ResponseCodeMapper(code = 500, type = TestRs500::class)
              @HttpRoute(method = "GET", path = "/test")
              fun test(): TestResponse
            }
        """.trimIndent(), """
            class TestMapper200 : HttpClientResponseMapper<TestRs200, Mono<TestRs200>> {
              override fun apply(rs: HttpClientResponse) = Mono.just(TestRs200(""))!!
            }
            class TestMapper500 : HttpClientResponseMapper<TestRs500, Mono<TestRs500>> {
              override fun apply(rs: HttpClientResponse) = Mono.just(TestRs500(""))!!
            }
        """.trimIndent(), """
            sealed interface TestResponse
            data class TestRs200(val test: String) : TestResponse
            data class TestRs500(val test: String) : TestResponse
        """.trimIndent())
        var result = client.invoke("test")
        assertThat(result).isEqualTo(new("TestRs200", ""))

        whenever(httpResponse.code()).thenReturn(500)
        result = client.invoke("test")
        assertThat(result).isEqualTo(new("TestRs500", ""))

        whenever(httpResponse.code()).thenReturn(201)
        assertThatThrownBy { client.invoke("test") }.isInstanceOf(HttpClientResponseException::class.java)
    }
}
